packages/website/src/components/RulesTable/index.tsx

```diff
@@ -224,3 +203,98 @@ export default function RulesTable({
     </>
   );
 }
+
+type FilterCategory =
+  | 'recommended'
+  | 'strict'
+  | 'fixable'
+  | 'suggestions'
+  | 'typeInformation';
+type FiltersState = Record<FilterCategory, FilterMode>;
+const neutralFiltersState: FiltersState = {
+  recommended: 'neutral',
+  strict: 'neutral',
+  fixable: 'neutral',
+  suggestions: 'neutral',
+  typeInformation: 'neutral',
+};
+
+function useRulesFilters(
+  paramsKey: string,
+): [FiltersState, (category: FilterCategory, mode: FilterMode) => void] {
+  const [state, setState] = useState(neutralFiltersState);
+
+  useIsomorphicLayoutEffect(() => {
+    const search = new URLSearchParams(window.location.search);
```

Comment:
You can replace `window.location` with `useLocation()` so you get SSR safety for free

Reply:
I expected `useLocation()` to trigger re-renders whenever the location changes, as the docs would suggest: https://reactrouter.com/en/main/hooks/use-location

So using `location` from `useLocation` here wouldn't be the same logic, right?
And I'm in a `useLayoutEffect` which does not run on the server, so accessing `window.location` works.

Could you tell me more what do you mean by "SSR safety for free"?

Notice that I had rewritten this logic to `react-router` specific hooks, what I described in **"Alternative solution"** in the original post. It turned out to be buggy and harder to reason about than the less elegant, naive solution with `setState` and `useLayoutEffect`, so I reverted that change — https://github.com/typescript-eslint/typescript-eslint/pull/6568/commits/786557fb52e5a18664f8291b853182a8cfd22a38.

> The implementation looks mostly good to me, but as some extra reading, you may want to try useSyncExternalStore ([interesting read](https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api)) ^^

Nice article! Do you think I should rewrite this PR to use `useHistorySelector` presented in the post?

Reply:
I'm okay either way! I don't have strong opinions on this, and TBH my React is getting a bit rusty these days.

Reply:
> I mean you don't have to use useIsomorphicLayoutEffect and read from a client-only API inside. useLocation is available both server-side and client-side, so it's more robust.

Well, yes, that's true, but then we shouldn't use `useState`, because a change to the location will already re-render the component. Would you agree?

Would you prefer if I went back to the state from https://github.com/typescript-eslint/typescript-eslint/pull/6568/commits/7c2e81a4f61fcaf7cbb2559ada50ef0b2d3e9b2b? and write end-to-end tests to ensure it works?

> I'm okay either way!

I won't be able to contribute any code here this week, so while I agree the solution would be more elegant with `useHistorySelector`, this PR would need to wait.